nucliadb_vectors/src/relations/bfs_engine.rs

```diff
@@ -76,55 +81,69 @@ where Guide: BfsGuide
 impl<'a, Guide> BfsEngine<'a, Guide>
 where Guide: BfsGuide
 {
-    pub fn search(mut self) -> RResult<Vec<Entity>> {
-        std::mem::take(&mut self.entry_points)
+    pub fn search(mut self) -> RResult<impl Iterator<Item = GCnx>> {
+        self.entry_points
             .iter()
             .copied()
             .map(|point| (BfsNode { point, depth: 0 }, self.visited.insert(point)))
             .filter(|(_, v)| *v)
             .for_each(|(e, _)| self.work_stack.push_back(e));
-        let mut results = vec![];
         while let Some(node) = self.work_stack.pop_front() {
             self.expand(node)?;
-            if self.guide.matches(node.point) {
-                results.push(node.point);
-            }
         }
-        Ok(results)
+        Ok(self.subgraph.into_iter())
     }
     fn expand(&mut self, node: BfsNode) -> RResult<()> {
+        // same_level nodes are reached by a free_edge
+        // which means that they belong to the level being explored now.
+        let mut same_level = HashMap::new();
+        // next_level nodes are reached by a edge that increases the level.
+        let mut next_level = HashMap::new();
         self.graph
             .get_outedges(self.txn, node.point)?
+            .chain(self.graph.get_inedges(self.txn, node.point)?)
             .flat_map(|a| a.ok().into_iter())
             .filter(|edge| node.depth < self.max_depth || self.guide.free_jump(*edge))
             .filter(|edge| self.guide.edge_allowed(edge.edge()))
             .filter(|edge| self.guide.node_allowed(edge.to()))
             .for_each(|edge| {
-                if !self.visited.contains(&edge.to()) {
+                let is_free_jump = self.guide.free_jump(edge);
+                let can_use_free_jump = same_level.contains_key(&node.point);
+                if !is_free_jump && !can_use_free_jump {
                     let node = BfsNode {
                         point: edge.to(),
-                        depth: node.depth + (!self.guide.free_jump(edge) as usize),
+                        depth: node.depth + 1,
```

Comment:
Can you explain that `+ 1` offset ?

Reply:
I think it's increasing the depth of the BFS, if it's not a free jump (i.e., it count as a jump), the next node is at the distance of the current node + 1

Reply:
Could you please add a comment to explain it ? I am sure @hermeGarcia will complete the command later on with further explanation :cat:

Reply:
Done